Sporadic test failure for mock source in no-drain mode

On very rare occasions one of the tests for the ring buffer source fails (locally, never happened in a CI pipeline so far):
```
	=== RUN   TestClosedSourceNoDrain
		./github.com/fako1024/slimcap/capture/afpacket/afring/afring_mock_test.go:139:
				Error Trace:	./github.com/fako1024/slimcap/capture/afpacket/afring/afring_mock_test.go:139
				Error:      	Expected nil, but got: capture.Packet{0x3, 0xe, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x40, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6, 0x0, 0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x15, 0xb3, 0x0, 0x50, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}
				Test:       	TestClosedSourceNoDrain
	--- FAIL: TestClosedSourceNoDrain (0.03s)
	=== RUN   TestFillRingBuffer
```
This happens _after_ closing the mock source, where of course we'd expect no further packets to arrive:
```
	// Close it right away
	require.Nil(t, mockSrc.Close())

	// Attempt to read from the source
	pkt, err := mockSrc.NextPacket(nil)
	require.Nil(t, pkt)
```
Most likely there is a bug / race somewhere in the `Close()` procedure that causes the source to still be open for a brief moment after it returns. Certainly not a critical issue (since it most likely only affects the mock source and doesn't really do any harm), but annoying still.

--- a/capture/afpacket/afring/afring_mock.go
+++ b/capture/afpacket/afring/afring_mock.go
@@ -176,7 +176,7 @@ func (m *MockSource) CanAddPackets() bool {
 // Pipe continuously pipes packets from the provided source through this one, mimicking
 // the ring buffer / TPacketHeader block retirement setting for population of the ring buffer
 func (m *MockSource) Pipe(src capture.Source, doneReadingChan chan struct{}) (errChan chan error) {
-	errChan = make(chan error)
+	errChan = make(chan error, 1)
 
 	go func(errs chan error, done chan struct{}) {
 		for {
@@ -220,9 +220,13 @@ func (m *MockSource) Done() {
 // ForceBlockRelease releases all blocks to the kernel (in order to "unblock" any potential mock capture
 // from the consuming routine without having to attempt a failed packet consumption)
 func (m *MockSource) ForceBlockRelease() {
-	for i := 0; i < m.nBlocks; i++ {
-		m.markBlock(i, unix.TP_STATUS_KERNEL)
-	}
+	m.markAllBlocks(unix.TP_STATUS_KERNEL)
+}
+
+// ForceBlockskUnavailable marks all blocks to be unavailable (in order to prevent any access to blocks
+// potentially unconsumed at / after a point in time where this method is called)
+func (m *MockSource) ForceBlockskUnavailable() {
+	m.markAllBlocks(unix.TP_STATUS_CSUMNOTREADY)
 }
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////
@@ -256,6 +260,12 @@ func (m *MockSource) getBlockStatus(n int) (status uint32) {
 
 func (m *MockSource) markBlock(n int, status uint32) {
 	atomic.StoreUint32((*uint32)(unsafe.Pointer(&m.ringBuffer.ring[n*m.blockSize+8])), status)
+}
+
+func (m *MockSource) markAllBlocks(status uint32) {
+	for i := 0; i < m.nBlocks; i++ {
+		m.markBlock(i, status)
+	}
 }
 
 func (m *MockSource) hasUserlandBlock() bool {

--- a/capture/afpacket/afring/afring_mock_nodrain.go
+++ b/capture/afpacket/afring/afring_mock_nodrain.go
@@ -61,11 +61,14 @@ func (m *MockSourceNoDrain) Run(releaseInterval time.Duration) (<-chan error, er
 	m.FinalizeBlock(false)
 	m.MockFd.SetNoRelease(true)
 
-	errChan := make(chan error)
+	errChan := make(chan error, 1)
 	m.wgRunning.Add(1)
 	go func(errs chan error) {
 
-		defer close(errs)
+		defer func() {
+			close(errs)
+			m.wgRunning.Done()
+		}()
 
 		// Queue / trigger a single event equivalent to receiving a new block via the PPOLL syscall and
 		// instruct the mock socket to not release the semaphore. That way data can be consumed immediately
@@ -79,10 +82,9 @@ func (m *MockSourceNoDrain) Run(releaseInterval time.Duration) (<-chan error, er
 		for {
 			for i := 0; i < m.nBlocks; i++ {
 
-				// If the mock source is closing return
+				// If the mock source is closing, mark all blocks as unavailable and return
 				if m.closing.Load() {
-					m.wgRunning.Done()
-					errs <- nil
+					m.ForceBlockskUnavailable()
 					return
 				}
 

